content/blog/2025-12-16-bracket.md

```diff
@@ -0,0 +1,261 @@
+triv      ::= x | fixnum | #t | #f | empty 
+          |   (void) | (error uint8) | ascii-char-literal
+x         ::= name? | prim-f
+prim-f    ::= binop | unop
+binop     ::= * | + | - | eq? | < | <= | > | >=
+unop      ::= fixnum? | boolean? | empty? | void?
+          |   ascii-char? | error? | not
+uint8     ::= uint8?
+ascii-char-literal ::= ascii-char-literal?
+fixnum    ::= int61?
+```
+
+The key design challenge was bridging the semantic gap between these educational IRs. Racket's `exprs-lang-v7` uses a dynamically-typed, tagged object representation where values are **ptrs** (64-bit words with tag bits encoding type information), while Bril uses explicit `any` types with runtime type checking.
+
+### Implementation: The `bracket` Pass
+
+The compiler is implemented as a single pass called `bracket` that performs a complete translation from `proc-imp-cmf-lang-v7` to Bril JSON.
+
+#### Tagged Object Representation Translation
+
+The source language uses a sophisticated tagging scheme where:
+- **Fixnums** use 3 least-significant bits as a primary tag (`#b000`), with the remaining 61 bits holding the actual integer value shifted left by 3
+- **Non-fixnum immediates** use a 3-bit primary tag (`#b110`) plus 5 additional bits for secondary tags
+
+This representation enables efficient operations: for example, `(* 8 n)` directly represents fixnum `n`, allowing addition and subtraction without untagging since `8x + 8y = 8(x + y)`.
+
+Since Bril's core instruction set doesn't include bitwise operations like `bitwise-and`, `bitwise-ior`, `bitwise-xor`, and `arithmetic-shift-right`, I used GenAI to generate a **standard library** that implements these operations using only Bril's integer arithmetic and boolean logic:
+
+- **`@bit(x: int): int`** - Extracts the least significant bit
+- **`@bitand(a: int, b: int): int`** - Implements bitwise AND through loops
+- **`@bitor(a: int, b: int): int`** - Implements bitwise OR
+- **`@bitxor(a: int, b: int): int`** - Implements bitwise XOR
+- **`@ashr(x: int, n: int): int`** - Implements arithmetic shift right with sign extension
+
+I implemented the logic in `bracket` to recognize when a `binop` is a bitwise operation and generate the appropriate calls to these standard library functions. These functions are prepended to every compiled program.
+
+#### Control Flow Translation
+
+The source IR uses predicates (relops, boolean combinators) that must be converted to Bril's branch instructions. My `compile-pred` function:
+- Translates relops directly to Bril comparison operations
+- Handles `!=` by compiling as `(not (= op1 op2))`
+- Transforms `not`, `(true)`, and `(false)` appropriately
+- Threads effects through predicate computation
+- Handles nested `if` expressions by creating fresh labels
+
+#### Function Calls and Tail Calls
+
+The compiler distinguishes between tail calls and non-tail calls:
+- Tail calls in `main` untag the result (divide by 8) before printing
```

Comment:
Does this mean that prints can only occur in the main function? (There is no way to print anywhere else in the language?)